File: Worksheets/Worksheet1_Answers.m
%
% Suggested model answers to worksheet 1
%
% Ian Hawke, 14/14/08
%

% Clear all starting values.
clear;
close all;
clc;

% Coding question 1; matrices

disp('Question 1')

A1 = [1 2; 3 4]
A2 = [-3 2; 3 6]

disp('The transpose of the matrices are')
A1.'
A2.'

disp('The inverse of the matrices are')
inv(A1)
inv(A2)

% Coding question 2; norms and condition numbers

disp('Question 2')

v1 = [1;3;-1];
v2 = [1;-2];
v3 = [1;6;-3;1];

disp('The vector 1-norms are')
norm(v1,1)
norm(v2,1)
norm(v3,1)
disp('The vector 2-norms are')
norm(v1,2)
norm(v2,2)
norm(v3,2)
disp('The vector infinity-norms are')
norm(v1,inf)
norm(v2,inf)
norm(v3,inf)

disp('The matrix 1-norms are')
norm(A1,1)
norm(A2,1)
disp('The matrix infinity-norms are')
norm(A1,inf)
norm(A2,inf)

% Coding question 3; writing a function

disp('Question 3')

Worksheet1_MyConditionNumber(A1)
Worksheet1_MyConditionNumber(A2)

% Coding question 4; bisection method

disp('Question 4')

% Set up the tolerance in the width of the interval
tolerance = 1e-15;
% Set up the initial bracketing interval
x_lo = 0;
x_hi = 1;

% Set up the function and the initial function values.
f = @(x)(tan(x) - exp(-x));
f_lo = f(x_lo);
f_hi = f(x_hi);

% Check that we really bracket a root properly
if (f_lo * f_hi > 0)
    error('The initial interval does not properly bracket a root!')
end

% Set up the bisection loop.
x_guess = (x_lo + x_hi) / 2;
f_guess = f(x_guess);

% Loop until the root is bracketed within the tolerance.
while (x_hi - x_lo > tolerance)
   
    if (f_lo * f_guess < 0)
       % The root is between the guessed value and the lower bound
       x_hi = x_guess;
       f_hi = f_guess;
    else
       % The root is between the guessed value and the upper bound
       x_lo = x_guess;
       f_lo = f_guess;
    end
    
    % Set the new bisection estimate
    x_guess = (x_lo + x_hi) / 2;
    f_guess = f(x_guess);
    
end

disp(sprintf('The final value of x is \n  %18.12g \nwhere the value of f is \n %18.12g', ...
    x_guess, f_guess));

File: Worksheets/Worksheet1_Answers.py
# -*- coding: utf-8 -*-
"""
Example solutions for Worksheet 1

Ian Hawke
"""

def MatrixConditionCheck(A):
    import numpy as np

    MaxConditionNumber = 10 # This is absurdly low    
    ConditionNumber = np.linalg.cond(A)
    if ConditionNumber > MaxConditionNumber:
        print "Condition number of matrix\n", A, "\ntoo large (bigger than", MaxConditionNumber, ").\n"
    

import numpy as np

print "\nQuestion 1\n"

A1 = np.array([[1,2],[3,4]])
A1T = np.transpose(A1)
A1I = np.linalg.inv(A1)
print "The matrix\n", A1, "\nhas transpose\n", A1T, "\nand inverse\n", A1I

A2 = np.array([[-3,2],[3,6]])
A2T = np.transpose(A2)
A2I = np.linalg.inv(A2)
print "The matrix\n", A2, "\nhas transpose\n", A2T, "\nand inverse\n", A2I

print "\nQuestion 2\n"

v1 = np.array([1,3,-1])
v2 = np.array([1,-2])
v3 = np.array([1,6,-3,1])

print "The vector\n", v1, "\nhas norms", np.linalg.norm(v1,1), np.linalg.norm(v1,2), np.linalg.norm(v1,np.inf)
print "The vector\n", v2, "\nhas norms", np.linalg.norm(v2,1), np.linalg.norm(v2,2), np.linalg.norm(v2,np.inf)
print "The vector\n", v3, "\nhas norms", np.linalg.norm(v3,1), np.linalg.norm(v3,2), np.linalg.norm(v3,np.inf)

print "The matrix\n", A1, "\nhas norms", np.linalg.norm(A1,1), np.linalg.norm(A1,np.inf)
print "The matrix\n", A2, "\nhas norms", np.linalg.norm(A2,1), np.linalg.norm(A2,np.inf)

print "\nQuestion 3\n"

MatrixConditionCheck(A1)
MatrixConditionCheck(A2)

print "\nQuestion 4\n"

# Bisection algorithm

tolerance = 1e-15
# Define the function
f = lambda x: np.tan(x) - np.exp(-x)
# Define the interval
x_min = 0.0
x_max = 1.0
# Values at the ends of the domain
f_min = f(x_min)
f_max = f(x_max)
assert(f_min * f_max < 0.0)
# The loop
x_c = (x_min + x_max) / 2.0
f_c = f(x_c)
iteration = 0
while ((x_max - x_min > tolerance) and (np.abs(f_c) > tolerance) and (iteration < 100)):
    iteration = iteration+1    
    if f_min * f_c < 0.0:
        x_max = x_c
        f_max = f_c
    else:
        x_min = x_c
        f_min = f_c
    x_c = (x_min + x_max) / 2.0
    f_c = f(x_c)
#    print "Iteration ", iteration, " x ", x_c, " f ", f_c

print "The root is approximately ", x_c, " where f is ", f_c
